design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,306 @@
+# KubeVirt Feature Lifecycle
+
+## Summary
+
+KubeVirt requires a clear policy on how features are introduced,
+evaluated and finally graduated or removed.
+
+This proposal defines the steps and policies to follow in order
+to manage a feature and its lifecycle in KubeVirt.
+
+The proposal is focusing on introducing features in
+a stable API (CRD) version.
```

Comment:
I think we need some out-of-band control for something like this:
see: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#adding-unstable-features-to-stable-versions

*When adding a feature to an object which is already Stable, the new fields and new behaviors need to meet the Stable level requirements. If these cannot be met, then the new field cannot be added to the object.*

Reply:
This is a great link, Simone!

@EdDev do you mind familiarizing yourself with his section?

Reply:
Done

Reply:
Actually my concern was somehow different:
I'm mostly concerned on how we should express the the feature gates themself.

Currently we are tracking them under `/spec/configuration/developerConfiguration/featureGates` on the `KubeVirt` kind which is at `"apiVersion": "kubevirt.io/v1"`.

So we named them as `featureGates` but at the end they are part of a stable API and so at the end they are configuration knobs and altering/removing them over the time will be hard.

For the same reason, another project - Openshift, in [its dev-guide is explicitly stating](https://github.com/openshift/enhancements/blob/master/dev-guide/featuresets.md#how-do-feature-gates-differ-from-configuration):

> **How do Feature Gates differ from configuration?**
> It’s true that it’s possible to release a feature that’s hidden behind configuration. The biggest difference is that our configuration is one of our APIs. Configuration never goes away. Feature Gates can and do disappear by design.

so, from this perspective, something like `/spec/configuration/developerConfiguration/featureGates` on  `"apiVersion": "kubevirt.io/v1"` is clearly part of the API and so it's a configuration knob and not an out-of-band feature gate.

Openshift for instance is solving this introducing a specific CRD for FeatureGates:
```
apiVersion: config.openshift.io/v1
kind: FeatureGate
```
and handling them as instances there.

So I think that in this context we should also design how we can define feature gates.